core/src/smartcontracts/wasm/codec/src/lib.rs

```diff
@@ -0,0 +1,142 @@
+        ..(offset + len).try_into().expect(U32_TO_USIZE_ERROR_MES)];
+
+    let obj =
+        T::decode_all(&mut &bytes[len_size_bytes.try_into().expect(U32_TO_USIZE_ERROR_MES)..])
+            .map_err(|err| Trap::new(format!("Failed to decode object: {err}")))?;
+
+    dealloc_fn
+        .call(&mut context, (offset, len))
+        .map_err(|err| Trap::new(format!("Failed to call dealloc fn: {err}")))?; // TODO
+    Ok(obj)
+}
+
+/// Encode `obj` to the given `memory` with the given `alloc_fn` and `context`.
+///
+/// Returns the offset of the encoded object.
+pub fn encode_into_memory<T: Encode>(
+    obj: &T,
+    memory: &wasmtime::Memory,
+    alloc_fn: &wasmtime::TypedFunc<WasmUsize, WasmUsize>,
+    mut context: impl wasmtime::AsContextMut,
+) -> Result<WasmUsize, Trap> {
+    let bytes = encode_with_length_prefix(obj);
+
+    let len = bytes
+        .len()
+        .try_into()
+        .expect("Checked in `encode_with_length_prefix`");
+
+    let offset = alloc_fn.call(&mut context, len)?;
+    let offset_usize = offset
+        .try_into()
+        .expect("`u32` should always fit in `usize`");
+
+    memory
+        .write(&mut context, offset_usize, &bytes)
+        .map_err(|err| Trap::new(err.to_string()))?;
+
+    Ok(offset)
+}
+
+/// Encode the given object but also add it's length in front of it. This can be considered
+/// a custom encoding format.
+///
+/// Usually, to retrieve the encoded object both pointer and the length of the allocation
+/// are provided. However, due to the lack of support for multivalue return values in stable
+/// `WebAssembly` it's not possible to return two values from a wasm function without some
+/// shenanignas. In those cases, only one value is sent which is pointer to the allocation
+/// with the first element being the length of the encoded object following it.
+pub fn encode_with_length_prefix<T: Encode>(obj: &T) -> Vec<u8> {
+    // Compile-time size check
+    #[allow(clippy::let_unit_value)]
+    let _ = <SizeChecker<T> as DoesSizeFitInWasmUsize>::RESULT;
+
+    let len_size_bytes = core::mem::size_of::<WasmUsize>();
+
+    let mut r = Vec::with_capacity(len_size_bytes + obj.size_hint());
+
+    // Reserve space for length
+    r.resize(len_size_bytes, 0);
+    obj.encode_to(&mut r);
+
+    // Store length as byte array in front of encoding
+    let len =
+        &WasmUsize::try_from(r.len()).expect("Length is checked at compile time, this is a bug");
+    r[..len_size_bytes].copy_from_slice(&len.to_le_bytes());
+
+    r
+}
+
+/// Trait to check that type size fits in [`WasmUsize`] at compile time.
+trait DoesSizeFitInWasmUsize {
+    const RESULT: ();
+}
+
+/// Helper struct to check type size at compile time.
+struct SizeChecker<T>(std::marker::PhantomData<T>);
+
+impl<T> DoesSizeFitInWasmUsize for SizeChecker<T> {
+    const RESULT: () = assert!(
+        core::mem::size_of::<T>() < (u32::MAX as usize - core::mem::size_of::<WasmUsize>()),
+        "Size doesn't fit in 2^32"
+    );
+}
```

Comment:
```suggestion
/// Helper struct to check type size at compile time.
struct SizeChecker<T>(std::marker::PhantomData<T>);

impl<T> SizeChecker<T> {
    const RESULT: () = assert!(
        core::mem::size_of::<T>() < (u32::MAX as usize - core::mem::size_of::<WasmUsize>()),
        "Size doesn't fit in 2^32"
    );
}
```

make it associated const

Reply:
Nice, thanks